dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProgramEncoder.java

```diff
@@ -78,6 +81,32 @@ public BooleanFormula encodeFullProgram() {
                 encodeDependencies());
     }
 
+    public BooleanFormula encodeControlBarrier() {
+        BooleanFormulaManager bmgr = context.getBooleanFormulaManager();
+        BooleanFormula enc = bmgr.makeTrue();
+        Map<Integer, List<ControlBarrier>> groups = context.getTask().getProgram().getThreads().stream()
+                .filter(Thread::hasScope)
+                .collect(groupingBy(this::getWorkgroupId,
+                        flatMapping(t -> t.getEvents(ControlBarrier.class).stream(), toList())));
+
+        for (List<ControlBarrier> events : groups.values()) {
+            for (ControlBarrier e1 : events) {
+                Expression id1 = e1.getId();
+                BooleanFormula allCF = context.controlFlow(e1);
+                for (ControlBarrier e2 : events) {
+                    if (!e1.getThread().equals(e2.getThread())) {
+                        Expression id2 = e2.getId();
+                        BooleanFormula sameId = context.equal(context.encodeExpressionAt(id1, e1), context.encodeExpressionAt(id2, e2));
+                        BooleanFormula cf = bmgr.or(context.controlFlow(e2), bmgr.not(sameId));
+                        allCF = bmgr.and(allCF, cf);
+                    }
+                }
+                enc = bmgr.and(enc, bmgr.equivalence(allCF, context.execution(e1)));
+            }
+        }
+        return enc;
```

Comment:
I think it is correct as it is. It must enforce barrier execution if there is CF and all other barriers can be executed. But it is not symmetric because the condition should be satisfied for each barrier individually.

At some point we can try a more elaborated encoding (maybe with an integer variable counting executed barriers), but let's keep it simple for now.

Reply:
Ok ... now I see the asymmetry